src/aggregator/aggregator/generic_elem.go

```diff
@@ -282,55 +288,96 @@ func (e *GenericElem) Consume(
 		e.Unlock()
 		return false
 	}
-	idx := 0
-	for range e.values {
-		// Bail as soon as the timestamp is no later than the target time.
-		if !isEarlierThanFn(e.values[idx].startAtNanos, resolution, targetNanos) {
-			break
-		}
-		idx++
-	}
 	e.toConsume = e.toConsume[:0]
-	if idx > 0 {
-		// Shift remaining values to the left and shrink the values slice.
-		e.toConsume = append(e.toConsume, e.values[:idx]...)
-		n := copy(e.values[0:], e.values[idx:])
-		// Clear out the invalid items to avoid holding references to objects
-		// for reduced GC overhead..
-		for i := n; i < len(e.values); i++ {
-			e.values[i].Reset()
+
+	// Evaluate and GC expired items.
+	valuesForConsideration := e.values
+	e.values = e.values[:0]
+	for _, value := range valuesForConsideration {
+		if !isEarlierThanFn(value.startAtNanos, resolution, targetNanos) {
+			e.values = append(e.values, value)
+			continue
+		}
+		expired := true
+		if e.resendEnabled {
+			// If resend is enabled, we only expire if the value is now outside the buffer past. It is safe to expire
+			// since any metrics intended for this value are rejected for being too late.
+			expiredNanos := targetNanos - e.opts.BufferForPastTimedMetricFn()(resolution).Nanoseconds()
+			expired = value.startAtNanos < expiredNanos
+		}
+
+		// Modify the by value copy with whether it needs time flush and accumulate.
+		copiedValue := value
+		copiedValue.onConsumeExpired = expired
+		e.toConsume = append(e.toConsume, copiedValue)
+
+		if !expired {
+			// Keep item. Expired values are GC'd below after consuming.
+			e.values = append(e.values, value)
 		}
-		e.values = e.values[:n]
 	}
 	canCollect := len(e.values) == 0 && e.tombstoned
 	e.Unlock()
 
+	var (
+		cascadeDirty  bool
+		prevTimeNanos int64
+	)
 	// Process the aggregations that are ready for consumption.
 	for i := range e.toConsume {
+		expired := e.toConsume[i].onConsumeExpired
 		timeNanos := timestampNanosFn(e.toConsume[i].startAtNanos, resolution)
+		// seed the previous timestamp if this is first consumed value.
+		if prevTimeNanos == 0 {
+			for k := range e.consumedValues {
+				if k > prevTimeNanos && k < timeNanos {
+					prevTimeNanos = k
+				}
+			}
```

Comment:
nit: As briefly discussed, we could extract some of this `find in map` code to be encapsulated by the map itself using a custom type for the map, i.e.:
```go
type consumedValuesMap map[xtime.UnixNano][]transformation.Datapoint

func (m consumedValuesMap) prevTimeNanosBeforeTimeNanos(timeNanos int64) (int64, bool) {
  var prevTimeNanos int64
  for k := range e.consumedValues {
    if k > prevTimeNanos && k < timeNanos {
      prevTimeNanos = k
    }
  }
  return prevTimeNanos, prevTimeNanos != 0
}

```